hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/listener/ClientSmartListenerService.java

```diff
@@ -143,22 +138,13 @@ public void shutdown() {
 
     @Override
     public void memberAdded(final MembershipEvent membershipEvent) {
-        executionService.executeInternal(new Runnable() {
-            @Override
-            public void run() {
-                synchronized (listenerRegLock) {
-                    Member member = membershipEvent.getMember();
-                    members.add(member);
-                    for (ClientRegistrationKey registrationKey : registrations.keySet()) {
-                        try {
-                            invoke(registrationKey, member.getAddress());
-                        } catch (Exception e) {
-                            logger.warning("Listener " + registrationKey + " can not added to new member " + member);
-                        }
-                    }
-                }
+        synchronized (listenerRegLock) {
```

Comment:
Why is this lock needed? My spider senses go off when I see a global lock with such a huge amount of 'external' code. Would it not be a lot easier to get rid of the lock have a single dedicated thread that takes care of all logic; and actor like thread? This way all modifications go through that object and you don't need to deal that much with synchronization.